File: 백준/Gold/1043. 거짓말/거짓말.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Main {
    private static int[] parent;

    private static int find(int x) {
        if (parent[x] == x) return x;

        return parent[x] = find(parent[x]);
    }

    private static void union(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);

        if (rootX != rootY) {
            parent[rootY] = rootX;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        int insidersCount = Integer.parseInt(st.nextToken());

        parent = new int[n + 1];
        for (int i = 0; i <= n; i++) {
            parent[i] = i;
        }

        int[] insiders = new int[insidersCount];
        if (insidersCount > 0) {
            for (int i = 0; i < insidersCount; i++) {
                insiders[i] = Integer.parseInt(st.nextToken());
            }
        }

        List<List<Integer>> parties = new ArrayList<>();
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int partySize = Integer.parseInt(st.nextToken());
            List<Integer> party = new ArrayList<>();
            for (int j = 0; j < partySize; j++) {
                party.add(Integer.parseInt(st.nextToken()));
            }
            parties.add(party);
        }

        for (List<Integer> party : parties) {
            if (!party.isEmpty()) {
                int first = party.get(0);
                for (int j = 1; j < party.size(); j++) {
                    union(first, party.get(j));
                }
            }
        }

        int result = 0;
        for (List<Integer> party : parties) {
            boolean canLie = true;

            if (!party.isEmpty()) {
                int partyRoot = find(party.get(0));

                for (int insider : insiders) {
                    if (find(insider) == partyRoot) {
                        canLie = false;
                        break;
                    }
                }
            }

            if (canLie) result++;
        }

        System.out.println(result);
    }
}

File: 백준/Gold/1043. 거짓말/거짓말.kt
import java.io.BufferedReader
import java.io.InputStreamReader

private lateinit var parent: IntArray

private fun find(x: Int): Int {
    if (parent[x] == x) return x
    parent[x] = find(parent[x])
    return parent[x]
}

private fun union(x: Int, y: Int) {
    val rootX = find(x)
    val rootY = find(y)

    if (rootX != rootY) {
        parent[rootY] = rootX
    }
}

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val (n, m) = br.readLine().split(" ").map { it.toInt() }
    val insiderTokens = br.readLine().split(" ").map { it.toInt() }
    val insiderCount = insiderTokens[0]
    val insiders = if (insiderCount > 0) {
        insiderTokens.drop(1).take(insiderCount)
    } else {
        emptyList()
    }

    parent = IntArray(n + 1) { it }

    val parties = List(m) {
        val tokens = br.readLine().split(" ").map { it.toInt() }
        val partySize = tokens[0]
        tokens.drop(1).take(partySize)
    }

    parties.forEach { party ->
        if (party.isNotEmpty()) {
            val first = party[0]
            party.drop(1).forEach { person ->
                union(first, person)
            }
        }
    }

    val result = parties.count { party ->
        if (party.isEmpty()) {
            true
        } else {
            val partyRoot = find(party[0])
            insiders.none { find(it) == partyRoot }
        }
    }

    println(result)
}